### PR TITLE
[LI-HOTFIX] Do not throw exceptions when internal headers are set and…

### PR DESCRIPTION
**Pull request by viswamy**
[LI-HOTFIX] Do not throw exceptions when internal headers are set and the coordinated channel with the broker is using Message_v1(magic = 1)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java [hunk 1]
```diff
@@ -417,8 +420,18 @@ private Long appendWithOffset(long offset, boolean isControlRecord, long timesta
             if (timestamp < 0 && timestamp != RecordBatch.NO_TIMESTAMP)
                 throw new IllegalArgumentException("Invalid negative timestamp " + timestamp);
 
-            if (magic < RecordBatch.MAGIC_VALUE_V2 && headers != null && headers.length > 0)
-                throw new IllegalArgumentException("Magic v" + magic + " does not support record headers");
+            if (magic < RecordBatch.MAGIC_VALUE_V2 && headers != null && headers.length > 0) {
+                int nonInternalHeaders = 0;
+                for (Header header : headers) {
+                    if (!header.key().startsWith(INTERNAL_HEADER_PREFIX)) {
+                        nonInternalHeaders++;
```
Comment (radai-rosenblatt) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java]: micro optimization - you can throw right here

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java [hunk 2]
```diff
@@ -784,6 +786,40 @@ public void testIdempotenceWithOldMagic() throws InterruptedException {
         accum.append(tp1, 0L, key, value, Record.EMPTY_HEADERS, null, 0);
     }
 
+    @Test
+    public void testRecordHeadersWithV1() {
```
Comment (radai-rosenblatt) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java]: i would be happier if we have an integration test (with an embedded broker booted) to make sure this works. there may be code after the records builder that barfs

File: core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala [hunk 3]
```diff
@@ -41,6 +42,42 @@ class PlaintextProducerSendTest extends BaseProducerSendTest {
...
 
+  @Test
+  def testRecordHeaders(): Unit = {
+    val producerProps = new Properties()
+    producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    producerProps.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer")
+    producerProps.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.StringSerializer")
+    val producer = new KafkaProducer[String, String](producerProps)
+    var recordHeaders = new RecordHeaders()
+    recordHeaders.add("RecordHeaderKey", "RecordHeaderValue".getBytes)
```
Comment (jonlee2) [inline core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala]: Can you add a comment why this is invalid? Also maybe rename it to invalidRecordHeaders or unsupportedRecordHeaders.

Also need a todo to delete this test when v2 format is supported.

File: core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala [hunk 4]
```diff
@@ -41,6 +42,42 @@ class PlaintextProducerSendTest extends BaseProducerSendTest {
...
+    )
+    try {
+      producer.send(invalidRecord).get()
+      throw new IllegalStateException("The invalid record should have thrown an exception")
+    } catch {
+      // Ignore the exception because a non internal header was introduced
+      case ignored: IllegalArgumentException =>
+    }
+
+    recordHeaders = new RecordHeaders()
```
Comment (jonlee2) [inline core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala]: likewise, validRecordHeaders?

File: core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala (hunk 3 above)
Comment (viswamy) [reply core/src/test/scala/integration/kafka/api/PlaintextProducerSendTest.scala]: I have already added it at the beginning of MemoryRecordsBuilder.java -- We can revert the whole commit by using that as reference.

File: core/src/test/scala/integration/kafka/api/RecordHeaderProducerSendTest.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.Properties
+
+import kafka.api.BaseProducerSendTest
+import kafka.server.KafkaConfig
+import org.apache.kafka.clients.producer.{KafkaProducer, ProducerConfig, ProducerRecord}
+import org.apache.kafka.common.header.internals.RecordHeaders
+import org.junit.Test
+
+class RecordHeaderProducerSendTest extends BaseProducerSendTest {
```
Comment (jonlee2) [inline core/src/test/scala/integration/kafka/api/RecordHeaderProducerSendTest.scala:28]: Do we need this test? Seems redundant with the one above? What's the actual difference?

Comment (viswamy) [reply core/src/test/scala/integration/kafka/api/RecordHeaderProducerSendTest.scala]: This is an end to end test to make sure that headers won't cause any issues in other parts of the client/server code.

**Review by jonlee2 — APPROVED** [review]

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java [hunk 6]
```diff
@@ -401,6 +404,17 @@ private int writeLegacyCompressedWrapperHeader() {
         return writtenCompressed;
     }
 
+    // Validates and throws an exception if the headers contain non-internal record headers
+    private void validateHeaders(Header[] headers) {
+        if (magic < RecordBatch.MAGIC_VALUE_V2 && headers != null && headers.length > 0) {
+            for (Header header : headers) {
+                if (!header.key().startsWith(INTERNAL_HEADER_PREFIX)) {
+                    throw new IllegalArgumentException("Magic v" + magic + " does not support record headers");
```
Comment (radai-rosenblatt) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java]: nit - print the header key (to make it easier to find the code setting it)

**Review by radai-rosenblatt — APPROVED** [review]

**Pull request merged**